File: typescript/leetcodehard/nQueens.ts
/*
https://leetcode.com/problems/n-queens-ii/

status - completed
*/

import { logp } from "../utils/logger";

function totalNQueens(n: number): number {
  const cols = new Set<number>();
  const diag1 = new Set<number>();
  const diag2 = new Set<number>();
  let count = 0;

  function backtrack(row: number): void {
    if (row === n) {
      count++;
      return;
    }

    for (let col = 0; col < n; col++) {
      const d1 = row - col;
      const d2 = row + col;
      if (cols.has(col) || diag1.has(d1) || diag2.has(d2)) continue;

      cols.add(col);
      diag1.add(d1);
      diag2.add(d2);

      backtrack(row + 1);

      cols.delete(col);
      diag1.delete(d1);
      diag2.delete(d2);
    }
  }

  backtrack(0);
  return count;
}

logp(`input n=4 expected 2 actual ${totalNQueens(4)}`);
logp(`input n=1 expected 1 actual ${totalNQueens(1)}`);
logp(`input n=8 expected 92 actual ${totalNQueens(8)}`);

File: python/leetcodehard/n_queens.py
"""
https://leetcode.com/problems/n-queens/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def solveNQueens(self, n):
        results = []
        columns = set()
        diagonals = set()
        anti_diagonals = set()
        placement = []

        def backtrack(row):
            if row == n:
                board = []
                for col in placement:
                    row_str = "." * col + "Q" + "." * (n - col - 1)
                    board.append(row_str)
                results.append(board)
                return

            for col in range(n):
                if col in columns or (row - col) in diagonals or (row + col) in anti_diagonals:
                    continue

                columns.add(col)
                diagonals.add(row - col)
                anti_diagonals.add(row + col)
                placement.append(col)

                backtrack(row + 1)

                columns.remove(col)
                diagonals.remove(row - col)
                anti_diagonals.remove(row + col)
                placement.pop()

        backtrack(0)
        return results


if __name__ == "__main__":
    test_input_1 = 4
    output = Solution().solveNQueens(test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = 1
    output = Solution().solveNQueens(test_input_2)
    Solution().log.info("Output is %s", output)
